raiden_contracts/contracts/services/UserDeposit.sol

```diff
@@ -0,0 +1,142 @@
+pragma solidity ^0.5.2;
+
+import "raiden/Token.sol";
+import "raiden/Utils.sol";
+
+contract UserDepositContract is Utils {
+    string constant public contract_version = "0.1.0";
+    uint constant public withdraw_delay = 100;  // time before withdraw is allowed in blocks
+
+    // Token to be used for the deposit
+    Token public token;
+
+    // Keep track of balances
+    mapping(address => uint256) public balances;
+    mapping(address => WithdrawPlan) public withdraw_plans;
+
+    /*
+     *  Structs
+     */
+    struct WithdrawPlan {
+        uint256 amount;
+        uint256 earliest_withdraw;  // earliest block at which withdraw is allowed
+    }
+
+    /*
+     *  Events
+     */
+
+    event BalanceReduced(address indexed owner, uint newBalance);
+    event WithdrawPlanned(address indexed withdrawer, uint plannedBalance);
+
+    /*
+     *  Modifiers
+     */
+
+    modifier canTransfer() {
+        // TODO: allow only access from MSC and 1-n payment contract
+        require(true);
+        _;
+    }
+
+    /*
+     *  Constructor
+     */
+
+    /// @notice Set the default values for the smart contract
+    /// @param _token_address The address of the token to use for rewards
+    constructor(
+        address _token_address
+    )
+        public
+    {
+        require(_token_address != address(0x0));
+        require(contractExists(_token_address));
+
+        token = Token(_token_address);
+        // Check if the contract is indeed a token contract
+        require(token.totalSupply() > 0);
+    }
+
+    /// @notice Deposit tokens. Idempotent function that sets the
+    /// total_deposit of tokens of the beneficiary.
+    /// Can be called by anyone several times and on behalf of other accounts
+    /// @param beneficiary The account benefiting from the deposit
+    /// @param total_deposit The sum of tokens, that have been deposited
+    function deposit(address beneficiary, uint256 total_deposit) public
+    {
+        require(total_deposit > balances[beneficiary]);
+
+        // Calculate the actual amount of tokens that will be transferred
+        uint256 added_deposit = total_deposit - balances[beneficiary];
+
+        balances[beneficiary] += added_deposit;
+        require(token.transferFrom(msg.sender, address(this), added_deposit));
+    }
```

Comment:
Note - some more invariant checks will be needed, but I understand that the current version is in work and they will probably be added later.

Reply:
And I think we can even skip the safe failsafe_addition. We know that the total token supply is a uint256 (ERC20 uses `uint` to return `totalSupply`), so we can't get large enough numbers to overflow.

Reply:
Looked over this now and I don't think it needs other checks.
Regarding `balances[beneficiary] += added_deposit;`, even the overflow here is protected against by `require(total_deposit > balances[beneficiary]);` -> edge case would be `total_deposit == 2^256 - 1` -> after `balances[beneficiary] += added_deposit;`, we have `balances[beneficiary] == 2^256 - 1`, so it is fine, as the rest of the checks will be done in `token.transferFrom`

> Do we have to expect malicious behavior of the token contract or will they be checked and approved in some way?

In general in Raiden, we assume malicious behavior up to a point (meaning, we always have state checks before the transfer, as you also have here). However, in this case, the deployer of UDC will choose the token & make sure the implementation is not malicious.